Source/WebCore/page/text-extraction/TextExtraction.cpp

```diff
@@ -64,30 +81,126 @@ static inline HashMap<Ref<Text>, String> collectText(ContainerNode& container)
+    if (lastTextNode)
+        emitTextForLastNode();
+
+    return nodesAndText;
+}
+
+struct TraversalContext {
+    const TextAndSelectedRangeMap visibleText;
+    const std::optional<WebCore::FloatRect> rectInRootView;
+    unsigned onlyCollectTextAndLinksCount { 0 };
+
+    inline bool shouldIncludeNodeWithRect(const FloatRect& rect) const
+    {
+        return !rectInRootView || rectInRootView->intersects(rect);
+    }
+};
+
+static inline TextAndSelectedRangeMap collectText(Document& document)
+{
+    auto fullRange = makeRangeSelectingNodeContents(*document.body());
+    auto selection = document.selection().selection();
+    TextNodesAndText textBeforeRangedSelection;
+    TextNodesAndText textInRangedSelection;
+    TextNodesAndText textAfterRangedSelection;
+    [&] {
+        if (selection.isRange()) {
+            auto selectionStart = selection.start();
+            auto selectionEnd = selection.end();
+            auto rangeBeforeSelection = makeSimpleRange(fullRange.start, selectionStart);
+            auto selectionRange = makeSimpleRange(selectionStart, selectionEnd);
+            auto rangeAfterSelection = makeSimpleRange(selectionEnd, fullRange.end);
+            if (rangeBeforeSelection && selectionRange && rangeAfterSelection) {
+                textBeforeRangedSelection = collectText(*rangeBeforeSelection);
+                textInRangedSelection = collectText(*selectionRange);
+                textAfterRangedSelection = collectText(*rangeAfterSelection);
+                return;
+            }
+        }
+        // Fall back to collecting the full document.
+        textBeforeRangedSelection = collectText(fullRange);
+    }();
+
+    TextAndSelectedRangeMap result;
+    for (auto& [node, text] : textBeforeRangedSelection)
+        result.add(node.ptr(), TextAndSelectedRange { text, std::nullopt });
+
+    bool isFirstSelectedNode = true;
+    for (auto& [node, text] : textInRangedSelection) {
+        if (std::exchange(isFirstSelectedNode, false)) {
+            if (auto entry = result.find(node.ptr()); entry != result.end() && entry->key == node.ptr()) {
+                entry->value = std::make_pair(
+                    makeString(entry->value.first, text),
+                    CharacterRange { entry->value.first.length(), text.length() }
+                );
+                continue;
+            }
+        }
+        result.add(node.ptr(), TextAndSelectedRange { text, CharacterRange { 0, text.length() } });
+    }
+
+    bool isFirstNodeAfterSelection = true;
+    for (auto& [node, text] : textAfterRangedSelection) {
+        if (std::exchange(isFirstNodeAfterSelection, false)) {
+            if (auto entry = result.find(node.ptr()); entry != result.end() && entry->key == node.ptr()) {
+                entry->value.first = makeString(entry->value.first, text);
+                continue;
+            }
+        }
+        result.add(node.ptr(), TextAndSelectedRange { text, std::nullopt });
```

Comment:
Nit: Since I saw some `{ }` elsewhere in the PR and I suppose we're trending towards that direction nowadays.

```suggestion
        result.add(node.ptr(), TextAndSelectedRange { text, { } });
```

Reply:
Though feel free to ignore if you think it harms readability here

Reply:
Ditto line 134

Reply:
Good catch! Yes, I've been meaning to prefer `{ }` over `std::nullopt` going forward, but it looks like I missed a spot.